src/rubyfmt.rb

```diff
@@ -2457,11 +2458,84 @@ def on_heredoc_end(*args, &blk)
     super
   end
 
-  def fixup_tstring_content_for_double_quotes(string)
-    string.gsub!("\\\\", "__RUBYFMT_SAFE_QUAD")
-    string.gsub!("\\", "\\\\\\\\")
-    string.gsub!("\"", "\\\"")
-    string.gsub!("__RUBYFMT_SAFE_QUAD", "\\\\\\\\")
+
+  FOUR_SLASHES_IN_TSTRING = "\\\\\\\\"
+  LITERAL_DOUBLE_QUOTE_IN_TSTRING = "\""
+  TWO_SLASHES_IN_TSTRING = "\\\\"
+  ONE_SLASH_IN_TSTRING = "\\"
+
+  # Rubyfmt renders all string literals as double quotes for consistency's sake
+  # this method fixes up individual tstring content items to ensure that they
+  # represent the same string literal once reformatted. This method is not
+  # called on string literals that were already surrounded by double quotes,
+  # and so this formatting only applies to tstring_contents from strings that
+  # need to be manipulated to correctly reform as a double quoted string
+  #
+  def fixup_tstring_content_for_double_quotes(string, is_single_quote:, delimiter:)
```

Comment:
I don't see any obvious changes we could make to reduce the complexity of this method, but I wondered when working on this bug if eventually splitting it into an explicit “unescape all the things” step followed by a “escape the relevant things” step could give us two smaller complex things instead of one big one?

Reply:
do you think the high amount of comments and well named constants makes this more understandable?

paging @jcoglan for review thoughts on how we could de-complex this.

Reply:
So the naming does make it clear what's being done here, but the things that make me nervous about this code is that you're working on bare strings, and the order of transformations appears to matter, including the fact you set up placeholders in early steps that are then removed later.

I've no idea to what extent this is possible, and in ruby it might be more pain than it's worth, but I was wondering whether you could tokenise the string first, and work on transforming that token stream, rather than using `gsub` which has the effect of introducing characters that have special meaning to later operations.

This would remove the need for placeholders, since each transformation would leave anything it didn't recognise untouched. It might also facilitate doing some of the "nested" transformations where you set some placeholders, do another transformation, then return to what you were originally doing. This is the sort of things blocks are great at.

The only caveat I have is that ruby doesn't have pattern-matching and so the code for this method may end up being a nightmare to read.